### PR TITLE
More Precision Combinations For GroupNorm

### PR DESCRIPTION
**Pull request by alpha0422**
This PR includes some more I/O and parameters precision combinations for GroupNorm. This PR also fixes bugs with FP16 parameters during training.
Original crediet @nekorobov and @Zhuoyao1012, commit a9396c36.

### CODE REVIEW

File: apex/contrib/csrc/group_norm/group_norm_nhwc.cpp [hunk 1]
```diff
@@ -867,6 +867,20 @@ int main(int argc, char **argv) {
...
+    precision = PrecisionMode::FP32IOFP32W;
+  }
+  else if (use_bf16)
+  {
+    precision = PrecisionMode::BF16IOFP32W;
+  }
+  else
+  {
+    precision = PrecisionMode::FP16IOFP32W;
+  }
```
Comment (crcrpar) [inline apex/contrib/csrc/group_norm/group_norm_nhwc.cpp:882]: nit-picking as `precision` doesn't seem to be modified after this initialization w/ if-else
```suggestion
  const auto precision = [&]() -> PrecisionMode {
    if (use_fp32) {
      return PrecisionMode::FP32IOFP32W;
    } else if (use_bf16) {
      return PrecisionMode::BF16IOFP32W;
    } else {
      return PrecisionMode::FP16IOFP32W;
    }
  }();
```

File: apex/contrib/csrc/group_norm/group_norm_nhwc.h [hunk 2]
```diff
@@ -64,9 +64,15 @@ static inline __device__ void spin_wait_(int *barrier, int step, int expected) {
...
-  FP32,
+  FP32IOFP16W,
+  FP32IOBF16W,
+  FP32IOFP32W,
+  FP16IOFP16W,
+  FP16IOBF16W,
+  FP16IOFP32W,
+  BF16IOFP16W,
+  BF16IOBF16W,
+  BF16IOFP32W,
```
Comment (crcrpar) [inline apex/contrib/csrc/group_norm/group_norm_nhwc.h:75]: could you leave a comment what the first and the second dtype mean?

**Review by crcrpar — COMMENTED** [review]
regarding the commit authors, we could use "Co-authored-by" for clarity: https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

File: apex/contrib/csrc/group_norm/group_norm_nhwc.h (hunk 2 above)
Comment (alpha0422) [reply apex/contrib/csrc/group_norm/group_norm_nhwc.h:76]: Sure, updated.

**Pull request merged**